text/0000-slatepack.md

```diff
@@ -0,0 +1,391 @@
+
+  - Existing ed25519 public keys from the wallet are bech32 encoded with `slatepack` as the `human-readable part` to build a `SlatePack Address`
+
+  - A `SlatePack Address` can be decoded to its ed25519 public key which can then be mapped to an x25519 public key to be used for encryption
+
+Wallets should support the ability for a user to both use a static, reusable `SlatePack Address` or to generate a unique `SlatePack Address` for each transaction if they wish. By default wallets should generate a new `SlatePack Address` for the sender address for each transaction.
+
+ed25519 keys are bech32 encoded as `SlatePack Addresses` rather than x25519 keys because the mapping from ed25519 to x25519 is more straight forward (x25519 public keys do not carry a `v` coordinate so they can map to two possible ed25519 public keys- this is solvable but using the ed25519 as the first order key avoids a potentially complex solution).
+
+#### Key Generation
+
+_The derivation strategy for ed25519 keys used for a `SlatePack Address` in the context of grin-wallet remains open_
+
+#### Example `SlatePack Address`
+
+```
+slatepack1p4fuklglxqsgg602hu4c4jl4aunu5tynyf4lkg96ezh3jefzpy6swshp5x
+```
+
+## `SlatePack Message`
+
+A `SlatePackMessage` requires multiple layers of data and encoding.
+
+### Serialization
+
+Grin slates are serialized as first order JSON objects. Binary serialization is done on those JSON objects. Before SlatePack, users could use both binary and JSON serialization for asynchronous transactions.
+
+With the SlatePack standard, all asynchronous transactions serialize the slate JSON objects as binary. JSON serialization for synchronous transactions (Tor) is still used as before. The SlatePack standard serialization choices are only relevant for asynchronous transaction methods.
+
+The details for the binary serialization of the most recent slates at the time of this writing can be found in the [Slate V4 (Compact Slates)](https://github.com/mimblewimble/grin-rfcs/pull/49) RFC. Future variations in slate binary serialization should be referenced in an RFC and may require the update of this document.
+
+
+### Metadata
+
+Metadata is included with SlatePack messages to indicate how to handle the encryption if any for the slata data in addition to tracking versions. It can be expanded in future versions with new fields that are safe to include in plain text. These fields are neither encrypted nor are they cryptographically authenticated by the sender. `mode` indicates which `Metadata` fields are expected to follow.
+
+- `"slatepack": [Major, Minor]`
+  - Where `[Major, Minor]` are positive fixnum ints representing the SlatePack version used to build the `SlatePack Message`
+
+- `"mode": int`
+  - Where `int` is a positive fixnum int indicating the type of SlatePack message
+  - 0 == plain text
```

Comment:
In some cases a user may not wish to reveal any extra data from their wallet (like a slatepack address). Without allowing for plain text we potentially force an extra step for every possible tx: exchanging an address before being able to exchange the initial slate. I expect that plain text will be a niche case but can be useful if we expect slatepack to be able to handle all possible cases.

Reply:
But that would no longer be a standardized transaction process. A standardized transaction begins with the receiver providing a SlatePack address. If you make that optional, you've once again require users to make a decision when receiving

Reply:
I agree that it shouldn't be an option that a user has to think about when transacting. A party shouldn't be making the decision of: "do I want to send/receive this slate encrypted or not? if I do, I better include my --encryption flag and generate/ask for an address."

So to be clear, by following the default standardized transaction flow, all async slate exchanges will be encrypted. However, as proposed, it is possible for a power user to go out of their way to use a plain text slate for a particular usecase. 

While still following the slatepack standard, it may not necessarily follow the _default flow_ of the standard if that makes sense? The default slatepack standard has a default flow which is not _required_ to follow the default standard- this means there would be the _encrypted flow_ (default, requires no extra thought/effort) and _plain text flow_ (requires extra steps and flags to achieve).

The concern with requiring a new SlatePackAddress to be used for _every_ new transaction is that it requires new data to be revealed by the wallet (impacts of this are not explicit but implicitly any data is privacy risk) and it requires an extra step for _every_ transaction (address exchange). I'm also imagining some business scenarios where for internal auditability it might be cleaner to use plaintext slatepacks so there isn't a need to securely manage key access for partners to review incoming/outgoing tx data.

Maybe this optionality adds some unnecessary complexity here which is definitely worth considering. If not this distinction definitely needs to be made more clear in the RFC. Curious for more feedback from others here as well.

Reply:
As long as we aren't adding a CLI option for plaintext, I would be fine with it. I just don't want to make it easy for exchanges, pools, and other services to support those. They have a long history of making less than ideal implementation choices and then being resistant to change once implemented.

Reply:
> Curious for more feedback from others here as well.

My view first of all is that we have a lot of unknowns at this point, for example: How many users will end up being accessible over Tor? How many services will end up being accessible over Tor? How many transactions will require manual slate exchange regardless as parties are not synchronously online? How will manual (slate) exchange between users and services perform when it's being deployed at scale for the first time? Where are the friction points? What's the breakdown of exchanges and mining pools that will choose to delist, upgrade, or roll some other home grown transaction building method? It makes it hard for me to reason about the impact of our decisions.

Secondly, as was pointed out before, enforcing an address exchange will not improve the security of transaction building if this occurs over the *same communication channel*. It does come at the cost of adding needless communication trips and friction for transactions that are already handled over encrypted channels. Conversely, allowing addresses to be optional do not break the consistency of the flow, or come at any other cost. 

So as we're making quite a lot of changes in one go, I'd like us to err on the side of caution and design the fall-back method in our initial attempt to be **simple to implement, easy to use, resistant to failure, and supportive of as many different uses as possible**. If we see that the protocol is well received, and that there's a well argued motivation for enforcing addresses for all transactions, this can still be introduced as part of the mandatory network upgrade to v5.0.0.

Reply:
> allowing addresses to be optional do not break the consistency of the flow

Yes it does. Users have to make decisions when sending if address is optional. There's no longer a consistent flow.

Reply:
The flow is consistent (as it remains the same), but you are correct in that the user makes a choice about whether to include an address or not. While we can debate the impact of this choice, my position is that it is worth while to allow for this, and a safer route to go in the first iteration of this protocol.